docs/fn-2.0/README.md

```diff
@@ -0,0 +1,416 @@
+# Fn 2.0
+
+## NOTE: This is a proposal for updating Fn with the following goals:
+
+* Split HTTP routes from functions
+* Support CloudEvent messages as the native format throughout
+* Simplify everything
+* Modularize components
+* Namespacing
+
+## Removals
+
+* Default, HTTP and JSON formats will be removed
+* Cold functions will be removed
+* Async, queues are event sources now
+
+## Terminology
+
+Fn follows the terminology defined here: https://github.com/cloudevents/spec/blob/master/spec.md#notations-and-terminology
+
+As well as the following:
+
+### Event Sources
+
+Event sources are things that emit events and pass the event messages to Fn. Three common event sources are:
+
+* HTTP
+* Scheduler
+* Queue
+
+### Functions
+
+Functions are little programs that accept a defined input and output a defined output. They do not need to be aware of any other parts
+of the system and can run standalone; from the command line, etc. These can be written in any language and are packaged up as a container image. 
+
+* Functions read from STDIN, write to STDOUT and log to STDERR
+* Functions are packaged in container images
+* Input is a stream of CloudEvent messages in JSON format
+* Output is a stream of CloudEvent messages in JSON format
+  * With some additions to deal with things like error handling and protocol specific responses (eg: function wants to return a specific http status code if it was fired from an HTTP event source.)
+
+#### Function Error Handling
+
+Errors will be returned as a CloudEvent with an “error” attribute, no “data” attribute, describing the error. 
+
+```
+{
+  "cloud-events-version": "0.1",
+  "event-id": "6480da1a-5028-4301-acc3-fbae628207b3",
+  "source": "http://example.com/repomanager",
+  "event-type": "http-request",
+  "event-type-version": "v1.5",
+  "event-time": "2018-04-01T23:12:34Z",
+  "content-type": "application/json",
+  “error”: {
+    "message": "an error occurred, blah, blah",
+    "type": "Exception",
+    "trace": "stack trace"
+  },
+  “extensions”: {
+    “protocol”: {
+       “headers”: …
+       Etc, same as we do now for JSON format. 
+    }
+  }
+}
+```
+
+Error map:
+* message - required
+* type - type of error, eg IOError - optional
```

Comment:
i think this one is redundant because the `message` is the most verbose representation of the error type, and if the trace is present the `type` will stay unused.

Reply:
Lambda has these three items, they all seem useful to me.